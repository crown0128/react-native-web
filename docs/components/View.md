# View spec

`View` is a flexbox container and the fundamental building block for UI. It is
designed to be nested inside other `View`'s and to have 0-to-many children of
any type.

## PropTypes

All other props are transferred directly to the `element`.

+ `component`: `func` or `string` (default `'div'`)
+ `pointerEvents`: `oneOf('all', 'box-only', 'box-none', 'none')`
+ `style`: `ViewStylePropTypes`

## ViewStylePropTypes

+ BackgroundPropTypes
+ BorderThemePropTypes
+ LayoutPropTypes
+ `boxShadow`: `string`
+ `color`: `string`
+ `opacity`: `number`

## ViewDefaultStyle

Implements the default styles from
[facebook/css-layout](https://github.com/facebook/css-layout).

1. All the flex elements are oriented from top-to-bottom, left-to-right and do
   not shrink. This is how things are laid out using the default CSS settings
   and what you'd expect.

2. The most convenient way to express the relation between width and other
   box-model properties.

3. Everything is `display:flex` by default. All the behaviors of `block` and
   `inline-block` can be expressed in term of flex but not the opposite.

4. Everything is `position:relative`. This makes `position:absolute` target the
   direct parent and not some parent which is either relative or absolute. If
   you want to position an element relative to something else, you should move
   it in the DOM instead of relying of CSS. It also makes `top`, `left`,
   `right`, `bottom` do something when not specifying `position:absolute`.

```js
const ViewDefaultStyle = {
  alignItems: 'stretch', // 1
  borderWidth: 0,
  borderStyle: 'solid',
  boxSizing: 'border-box', // 2
  display: 'flex', // 3
  flexBasis: 'auto', // 1
  flexDirection: 'column', // 1
  flexShrink: 0, // 1
  listStyle: 'none',
  margin: 0,
  padding: 0,
  position: 'relative' // 4
};
```

## Examples

```js
// TODO
```
