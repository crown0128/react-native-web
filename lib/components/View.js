import {objectWithProps} from '../getOtherProps';
import React, {PropTypes} from 'react';
import StylePropTypes from '../StylePropTypes';
import SDKComponent from './SDKComponent';

// https://github.com/facebook/css-layout#default-values
const ViewStyleDefaultProps = {
  alignItems: 'stretch',
  borderWidth: 0,
  borderStyle: 'solid',
  boxSizing: 'border-box',
  display: 'flex',
  flexBasis: 'auto',
  flexDirection: 'column',
  flexShrink: 0,
  listStyle: 'none',
  margin: 0,
  padding: 0,
  position: 'relative'
};

const ViewStylePropTypes = {
  ...StylePropTypes.BackgroundPropTypes,
  ...StylePropTypes.BorderThemePropTypes,
  ...StylePropTypes.LayoutPropTypes,
  boxShadow: PropTypes.string,
  opacity: PropTypes.number,
  transform: PropTypes.string
};

class View extends React.Component {
  static _getPropTypes() {
    return {
      className: PropTypes.string,
      element: PropTypes.oneOfType([
        PropTypes.string, PropTypes.func
      ]),
      pointerEvents: PropTypes.oneOf([
        'auto',
        'box-none',
        'box-only',
        'none'
      ]),
      style: PropTypes.shape(ViewStylePropTypes)
    };
  }

  static _getDefaultProps() {
    return {
      className: '',
      element: 'div',
      style: {}
    };
  }

  render() {
    const { className, element, pointerEvents, style, ...other } = this.props;
    const filteredStyle = objectWithProps(style, Object.keys(ViewStylePropTypes));
    const pointerEventsStyle = pointerEvents && { pointerEvents };
    const mergedStyle = {
      ...ViewStyleDefaultProps,
      ...filteredStyle,
      ...pointerEventsStyle
    };

    return (
      <SDKComponent
        {...other}
        className={`View ${className}`}
        element={element}
        style={mergedStyle}
      />
    );
  }
}

View.propTypes = View._getPropTypes();
View.defaultProps = View._getDefaultProps();

export default View;
export {ViewStylePropTypes};
