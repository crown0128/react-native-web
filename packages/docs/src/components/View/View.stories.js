import PropTypes, { any, bool, func, object, oneOf, string } from 'prop-types';

const ofProps = () => {};

ofProps.propTypes = {
  accessibilityLabel: PropTypes.string,
  accessibilityLiveRegion: oneOf(['assertive', 'none', 'polite']),
  accessibilityRelationship: object,
  accessibilityRole: string,
  accessibilityState: object,
  accessibilityValue: object,
  accessible: bool,
  children: any,
  forwardedRef: any,
  hitSlop: object,
  importantForAccessibility: oneOf(['auto', 'no', 'no-hide-descendants', 'yes']),
  itemID: string,
  itemProp: string,
  itemRef: string,
  itemScope: string,
  itemType: string,
  nativeID: string,
  onBlur: func,
  onContextMenu: func,
  onFocus: func,
  onKeyDown: func,
  onKeyUp: func,
  onLayout: func,
  onMoveShouldSetResponder: func,
  onMoveShouldSetResponderCapture: func,
  onResponderGrant: func,
  onResponderMove: func,
  onResponderReject: func,
  onResponderRelease: func,
  onResponderTerminate: func,
  onResponderTerminationRequest: func,
  onStartShouldSetResponder: func,
  onStartShouldSetResponderCapture: func,
  onWheel: func,
  pointerEvents: oneOf(['auto', 'box-none', 'box-only', 'none']),
  style: object,
  testID: string
};

export default {
  title: 'Components|View',
  includeStories: []
};

export { ofProps };
export { default as onLayout } from './examples/OnLayout';
export { default as pointerEvents } from './examples/PointerEvents';
export { default as styleBorder } from './examples/StyleBorder';
export { default as styleBoxShadow } from './examples/StyleBoxShadow';
export { default as styleFlexbox } from './examples/StyleFlexbox';
export { default as styleZIndex } from './examples/StyleZIndex';
