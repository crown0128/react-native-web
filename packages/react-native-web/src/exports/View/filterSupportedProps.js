const whitelist = {
  accessibilityComponentType: true,
  accessibilityLabel: true,
  accessibilityLiveRegion: true,
  accessibilityRole: true,
  accessibilityTraits: true,
  accessible: true,
  children: true,
  disabled: true,
  importantForAccessibility: true,
  onBlur: true,
  onContextMenu: true,
  onFocus: true,
  onMoveShouldSetResponder: true,
  onMoveShouldSetResponderCapture: true,
  onResponderEnd: true,
  onResponderGrant: true,
  onResponderMove: true,
  onResponderReject: true,
  onResponderRelease: true,
  onResponderStart: true,
  onResponderTerminate: true,
  onResponderTerminationRequest: true,
  onScrollShouldSetResponder: true,
  onScrollShouldSetResponderCapture: true,
  onSelectionChangeShouldSetResponder: true,
  onSelectionChangeShouldSetResponderCapture: true,
  onStartShouldSetResponder: true,
  onStartShouldSetResponderCapture: true,
  onTouchCancel: true,
  onTouchCancelCapture: true,
  onTouchEnd: true,
  onTouchEndCapture: true,
  onTouchMove: true,
  onTouchMoveCapture: true,
  onTouchStart: true,
  onTouchStartCapture: true,
  pointerEvents: true,
  style: true,
  testID: true,
  // escape-hatches for ScrollView
  onScroll: true,
  onWheel: true,
  // escape-hatches for Touchable keyboard support
  onKeyDown: true,
  onKeyPress: true,
  onKeyUp: true,
  // escape-hatches for creating hover effects
  onMouseEnter: true,
  onMouseLeave: true,
  onMouseMove: true,
  onMouseOver: true,
  onMouseOut: true,
  // unstable escape-hatches for web
  className: true,
  href: true,
  onClick: true,
  onClickCapture: true,
  rel: true,
  target: true
};

const filterSupportedProps = props => {
  const safeProps = {};
  for (const prop in props) {
    if (props.hasOwnProperty(prop)) {
      if (whitelist[prop] || prop.indexOf('aria-') === 0 || prop.indexOf('data-') === 0) {
        safeProps[prop] = props[prop];
      }
    }
  }
  return safeProps;
};

export default filterSupportedProps;
