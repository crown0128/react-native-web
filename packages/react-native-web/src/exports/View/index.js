/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import type { ViewProps } from './types';

import createElement from '../createElement';
import css from '../StyleSheet/css';
import setAndForwardRef from '../../modules/setAndForwardRef';
import useElementLayout from '../../hooks/useElementLayout';
import usePlatformMethods from '../../hooks/usePlatformMethods';
import StyleSheet from '../StyleSheet';
import TextAncestorContext from '../Text/TextAncestorContext';
import React, { forwardRef, useContext, useRef } from 'react';

export type { ViewProps };

function createHitSlopElement(hitSlop) {
  const hitSlopStyle = {};
  for (const prop in hitSlop) {
    if (hitSlop.hasOwnProperty(prop)) {
      const value = hitSlop[prop];
      hitSlopStyle[prop] = value > 0 ? -1 * value : 0;
    }
  }
  return createElement('span', {
    classList: [classes.hitSlop],
    style: hitSlopStyle
  });
}

const View = forwardRef<ViewProps, *>((props, ref) => {
  const {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessibilityRelationship,
    accessibilityRole,
    accessibilityState,
    accessibilityValue,
    forwardedRef,
    hitSlop,
    importantForAccessibility,
    nativeID,
    onBlur,
    onContextMenu,
    onFocus,
    onLayout,
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,
    pointerEvents,
    testID,
    // unstable
    onClick,
    onClickCapture,
    onScroll,
    onWheel,
    onKeyDown,
    onKeyPress,
    onKeyUp,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onTouchCancel,
    onTouchCancelCapture,
    onTouchEnd,
    onTouchEndCapture,
    onTouchMove,
    onTouchMoveCapture,
    onTouchStart,
    onTouchStartCapture,
    href,
    itemID,
    itemRef,
    itemProp,
    itemScope,
    itemType,
    rel,
    target,
    unstable_ariaSet,
    unstable_dataSet
  } = props;

  if (process.env.NODE_ENV !== 'production') {
    React.Children.toArray(props.children).forEach(item => {
      if (typeof item === 'string') {
        console.error(`Unexpected text node: ${item}. A text node cannot be a child of a <View>.`);
      }
    });
  }

  const hasTextAncestor = useContext(TextAncestorContext);
  const hostRef = useRef(null);
  const setRef = setAndForwardRef({
    getForwardedRef: () => forwardedRef,
    setLocalRef: c => {
      hostRef.current = c;
    }
  });

  const children = hitSlop
    ? React.Children.toArray([createHitSlopElement(hitSlop), props.children])
    : props.children;
  const classList = [classes.view];
  const style = StyleSheet.compose(
    hasTextAncestor && styles.inline,
    props.style
  );

  useElementLayout(hostRef, onLayout);
  usePlatformMethods(hostRef, ref, classList, style);

  return createElement('div', {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessibilityRelationship,
    accessibilityRole,
    accessibilityState,
    accessibilityValue,
    children,
    classList,
    importantForAccessibility,
    nativeID,
    onBlur,
    onContextMenu,
    onFocus,
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,
    pointerEvents,
    ref: setRef,
    style,
    testID,
    // unstable
    onClick,
    onClickCapture,
    onScroll,
    onWheel,
    onKeyDown,
    onKeyPress,
    onKeyUp,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onTouchCancel,
    onTouchCancelCapture,
    onTouchEnd,
    onTouchEndCapture,
    onTouchMove,
    onTouchMoveCapture,
    onTouchStart,
    onTouchStartCapture,
    href,
    itemID,
    itemRef,
    itemProp,
    itemScope,
    itemType,
    rel,
    target,
    unstable_ariaSet,
    unstable_dataSet
  });
});

View.displayName = 'View';

const classes = css.create({
  view: {
    alignItems: 'stretch',
    border: '0 solid black',
    boxSizing: 'border-box',
    display: 'flex',
    flexBasis: 'auto',
    flexDirection: 'column',
    flexShrink: 0,
    margin: 0,
    minHeight: 0,
    minWidth: 0,
    padding: 0,
    position: 'relative',
    zIndex: 0
  },
  // this zIndex-ordering positions the hitSlop above the View but behind
  // its children
  hitSlop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: -1
  }
});

const styles = StyleSheet.create({
  inline: {
    display: 'inline-flex'
  }
});

export default View;
