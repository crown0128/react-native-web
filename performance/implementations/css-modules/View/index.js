/* eslint-disable react/prop-types */
import classnames from 'classnames';
import React from 'react';
import styles from './styles.css';

const View = props => <div {...props} className={classnames(styles.initial, props.className)} />;

module.exports = View;
