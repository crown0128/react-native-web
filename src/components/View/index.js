import { pickProps } from '../../modules/filterObjectProps'
import CoreComponent from '../CoreComponent'
import React, { PropTypes } from 'react'
import ViewStylePropTypes from './ViewStylePropTypes'

const viewStyleKeys = Object.keys(ViewStylePropTypes)

const styles = {
  // https://github.com/facebook/css-layout#default-values
  initial: {
    alignItems: 'stretch',
    borderWidth: 0,
    borderStyle: 'solid',
    boxSizing: 'border-box',
    display: 'flex',
    flexBasis: 'auto',
    flexDirection: 'column',
    flexShrink: 0,
    listStyle: 'none',
    margin: 0,
    padding: 0,
    position: 'relative',
    // button reset
    backgroundColor: 'transparent',
    color: 'inherit',
    font: 'inherit',
    textAlign: 'inherit'
  }
}

class View extends React.Component {
  static propTypes = {
    _className: PropTypes.string, // escape-hatch for code migrations
    accessibilityLabel: PropTypes.string,
    accessibilityLiveRegion: PropTypes.oneOf(['assertive', 'off', 'polite']),
    accessibilityRole: PropTypes.string,
    accessible: PropTypes.bool,
    children: PropTypes.any,
    component: CoreComponent.propTypes.component,
    pointerEvents: PropTypes.oneOf(['auto', 'box-none', 'box-only', 'none']),
    style: PropTypes.shape(ViewStylePropTypes),
    testID: CoreComponent.propTypes.testID
  }

  static stylePropTypes = ViewStylePropTypes

  static defaultProps = {
    _className: '',
    accessible: true,
    component: 'div',
    style: styles.initial
  }

  render() {
    const {
      _className,
      accessibilityLabel,
      accessibilityLiveRegion,
      accessibilityRole,
      accessible,
      pointerEvents,
      style,
      testID,
      ...other
    } = this.props

    const className = `View ${_className}`.trim()
    const pointerEventsStyle = pointerEvents && { pointerEvents }
    const resolvedStyle = pickProps(style, viewStyleKeys)

    return (
      <CoreComponent
        {...other}
        aria-hidden={accessible ? null : true}
        aria-label={accessibilityLabel}
        aria-live={accessibilityLiveRegion}
        className={className}
        role={accessibilityRole}
        style={{
          ...styles.initial,
          ...resolvedStyle,
          ...pointerEventsStyle
        }}
        testID={testID}
      />
    )
  }
}

export default View
