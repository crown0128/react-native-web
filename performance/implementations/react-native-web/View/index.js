import View from 'react-native/components/View';
export default View;
