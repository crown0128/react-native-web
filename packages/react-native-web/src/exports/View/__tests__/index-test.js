/* eslint-env jasmine, jest */

import React from 'react';
import View from '../';
import { render } from '@testing-library/react';

describe('components/View', () => {
  test('default', () => {
    const { container } = render(<View />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('non-text is rendered', () => {
    const children = <View testID="1" />;
    const { container } = render(<View>{children}</View>);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('raw text nodes as children', () => {
    beforeEach(() => {
      jest.spyOn(console, 'error');
      console.error.mockImplementation(() => {});
    });

    afterEach(() => {
      console.error.mockRestore();
    });

    test('error logged (single)', () => {
      render(<View>hello</View>);
      expect(console.error).toBeCalled();
    });

    test('error logged (array)', () => {
      render(
        <View>
          <View />
          hello
          <View />
        </View>
      );
      expect(console.error).toBeCalled();
    });
  });

  describe('prop "hitSlop"', () => {
    test('renders a span with negative position offsets', () => {
      const { container } = render(<View hitSlop={{ top: 10, bottom: 10, right: 5, left: 5 }} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('handles partial offsets', () => {
      const { container } = render(<View hitSlop={{ top: 10 }} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('prop "pointerEvents"', () => {
    const { container } = render(<View pointerEvents="box-only" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
