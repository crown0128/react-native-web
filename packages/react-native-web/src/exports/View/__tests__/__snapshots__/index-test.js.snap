// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/View allows "dir" to be overridden 1`] = `
<div
  class="css-view-1dbjc4n"
/>
`;

exports[`components/View default 1`] = `
<div
  class="css-view-1dbjc4n"
/>
`;

exports[`components/View non-text is rendered 1`] = `
<div
  class="css-view-1dbjc4n"
>
  <div
    class="css-view-1dbjc4n"
    data-testid="1"
  />
</div>
`;

exports[`components/View prop "accessibilityLabel" value is set 1`] = `
<div
  aria-label="accessibility label"
  class="css-view-1dbjc4n"
/>
`;

exports[`components/View prop "accessibilityLiveRegion" value is set 1`] = `
<div
  aria-live="polite"
  class="css-view-1dbjc4n"
/>
`;

exports[`components/View prop "accessibilityRole" value alters HTML element 1`] = `
<a
  class="css-reset-4rbku5 css-cursor-18t94o4 css-view-1dbjc4n"
  role="link"
/>
`;

exports[`components/View prop "accessibilityRole" value is "button" 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n"
  role="button"
  tabindex="0"
/>
`;

exports[`components/View prop "accessibilityRole" value is set 1`] = `
<div
  class="css-view-1dbjc4n"
  role="presentation"
/>
`;

exports[`components/View prop "nativeID" value is set 1`] = `
<div
  class="css-view-1dbjc4n"
  id="nativeID"
/>
`;

exports[`components/View prop "pointerEvents" 1`] = `
<div
  class="css-view-1dbjc4n r-pointerEvents-ah5dr5"
/>
`;

exports[`components/View prop "style" value is set 1`] = `
<div
  class="css-view-1dbjc4n"
  style="border-top-width: 5px; border-right-width: 5px; border-bottom-width: 5px; border-left-width: 5px;"
/>
`;

exports[`components/View prop "testID" value is set 1`] = `
<div
  class="css-view-1dbjc4n"
  data-testid="123"
/>
`;
